### PR TITLE
register factory methods for proper garbage collection

### PR DESCRIPTION
**Pull request by shubham808**
#4321

### CODE REVIEW

File: src/interfaces/swig/factory.i [hunk 1]
```diff
@@ -4,3 +4,18 @@
 %include <shogun/util/factory.h>
 
 %template(features) shogun::features<float64_t>;
+
+/* These functions return a new Object */
+%newobject distance(const std::string& name);
+%newobject evaluation(const std::string& name);
+%newobject kernel(const std::string& name);
+%newobject kernel(SGMatrix<T> kernel_matrix);
```
[the remark this answers is not in the document]
Comment (shubham808) [reply src/interfaces/swig/factory.i]: mmh... i thought that too but did not find any @vigsterkr any ideas ?

[the remark this answers is not in the document]
Comment (vigsterkr) [reply src/interfaces/swig/factory.i]: as far as SWIG goes it doesn't understand templates, before they are actually instantiated, say

%template SGMatrix<float32_t>
%template SGMatrix<float64_t>

that's the general guideline of SWIG... dunno how newobject works in this context but i would expect the same (consistent) behaviour?

File: src/interfaces/swig/factory.i [hunk 2]
```diff
@@ -4,3 +4,18 @@
...
+%newobject machine(const std::string& name);
+%newobject multiclass_strategy(const std::string& name);
+%newobject ecoc_encoder(const std::string& name);
+%newobject ecoc_decoder(const std::string& name);
+%newobject features(SGMatrix<T> mat);
+%newobject features(CFile* file, EPrimitiveType primitive_type);
```
Comment (karlnapf) [inline src/interfaces/swig/factory.i]: in fact we dont even need the parameter names, just the types

File: src/interfaces/swig/factory.i (hunk 1 above)
[the remark this answers is not in the document]
Comment (karlnapf) [reply src/interfaces/swig/factory.i]: just checked, and there is no case of `%newobject` in shogun that is templated.
@shubham808 i think let's just do it for float64 for now ...

**Comment by karlnapf** [comment]
Ok looks good now.

Now what does the garbage collector say in SWIG interfaces? works?

**Comment by shubham808** [comment]
i am not sure...

**Comment by karlnapf** [comment]
It is pretty simple to find this out using valgrind ...

**Comment by shubham808** [comment]
valgrind says 0 leaks for all meta examples in python

**Comment by vigsterkr** [comment]
Have u made sure that all the hacks are being in place for memleak detection of python tool with valgrind, eg https://stackoverflow.com/a/45248015

**Comment by karlnapf** [comment]
Key is the delta of this before and after patch

**Comment by shubham808** [comment]
i used the gcdebug log level in a python script that overwrites a kernel generated from factory.
The object is not being destroyed. :(

**Comment by karlnapf** [comment]
Yeah ok, we will need to dig into this a bit I guess.....I wonder whether it is the SG_REF thingi in the factory that causes it....

**Comment by karlnapf** [comment]
I leave this open for now as we should fix all the memory leaks in the examples and the factory at some point

**Comment by vigsterkr** [comment]
"superseded" by #4604

**Pull request closed without merging**